File: bl-central-server/bl-web-app/src/main/java/org/jesperancinha/logistics/web/BridgeLogisticsInitializer.java
package org.jesperancinha.logistics.web;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.introspect.AnnotatedMember;
import com.fasterxml.jackson.databind.introspect.AnnotatedParameter;
import com.fasterxml.jackson.databind.introspect.JacksonAnnotationIntrospector;
import lombok.extern.slf4j.Slf4j;
import org.jesperancinha.logistics.jpa.dao.Bridge;
import org.jesperancinha.logistics.jpa.dao.BridgeOpeningTime;
import org.jesperancinha.logistics.jpa.dao.Carriage;
import org.jesperancinha.logistics.jpa.dao.Company;
import org.jesperancinha.logistics.jpa.dao.Container;
import org.jesperancinha.logistics.jpa.dao.Freight;
import org.jesperancinha.logistics.jpa.dao.Passenger;
import org.jesperancinha.logistics.jpa.dao.Product;
import org.jesperancinha.logistics.jpa.dao.Train;
import org.jesperancinha.logistics.jpa.dao.BridgeRepository;
import org.jesperancinha.logistics.jpa.dao.CarriageRepository;
import org.jesperancinha.logistics.jpa.dao.CompanyRepository;
import org.jesperancinha.logistics.jpa.dao.ContainerRepository;
import org.jesperancinha.logistics.jpa.dao.FreightRepository;
import org.jesperancinha.logistics.jpa.dao.OpeningTimeRepository;
import org.jesperancinha.logistics.jpa.dao.PassengerRepository;
import org.jesperancinha.logistics.jpa.dao.ProductRepository;
import org.jesperancinha.logistics.jpa.dao.TrainRepository;
import org.jesperancinha.logistics.web.data.FreightDto;
import org.jesperancinha.logistics.web.data.TrainDto;
import org.springframework.boot.CommandLineRunner;
import org.springframework.context.annotation.Profile;
import org.springframework.stereotype.Component;

import java.time.Instant;
import java.util.Arrays;
import java.util.Objects;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import static java.util.Arrays.stream;

@Component
@Slf4j
@Profile({"local", "demo"})
public class BridgeLogisticsInitializer implements CommandLineRunner {

    private final BridgeRepository bridgeRepository;

    private final CarriageRepository carriageRepository;

    private final ContainerRepository containerRepository;

    private final FreightRepository freightRepository;

    private final ProductRepository productRepository;

    private final TrainRepository trainRepository;

    private final OpeningTimeRepository openingTimeRepository;

    private final CompanyRepository companyRepository;

    private final PassengerRepository passengerRepository;

    private final ObjectMapper objectMapper = new ObjectMapper();

    public BridgeLogisticsInitializer(BridgeRepository bridgeRepository, CarriageRepository carriageRepository, ContainerRepository containerRepository, FreightRepository freightRepository, ProductRepository productRepository, TrainRepository trainRepository,
                                      OpeningTimeRepository openingTimeRepository, CompanyRepository companyRepository, PassengerRepository passengerRepository) {
        this.bridgeRepository = bridgeRepository;
        this.carriageRepository = carriageRepository;
        this.containerRepository = containerRepository;
        this.freightRepository = freightRepository;
        this.productRepository = productRepository;
        this.trainRepository = trainRepository;
        this.openingTimeRepository = openingTimeRepository;
        this.companyRepository = companyRepository;
        this.passengerRepository = passengerRepository;
        JacksonAnnotationIntrospector implicitRecordAI = new JacksonAnnotationIntrospector() {
            @Override
            public String findImplicitPropertyName(AnnotatedMember m) {
                if (m.getDeclaringClass()
                        .isRecord()) {
                    if (m instanceof AnnotatedParameter parameter) {
                        return m.getDeclaringClass()
                                .getRecordComponents()[parameter.getIndex()].getName();
                    }
                }
                return super.findImplicitPropertyName(m);
            }
        };
        objectMapper.setAnnotationIntrospector(implicitRecordAI);
    }

    @Override
    public void run(String... args) throws Exception {
        bridgeRepository.saveAll(Arrays.asList(objectMapper.readValue(getClass().getResourceAsStream("/bridges.json"), Bridge[].class)));

        companyRepository.saveAll(Arrays.asList(objectMapper.readValue(getClass().getResourceAsStream("/companies.json"), Company[].class)));

        carriageRepository.saveAll(Arrays.asList(objectMapper.readValue(getClass().getResourceAsStream("/carriages.json"), Carriage[].class)));

        containerRepository.saveAll(Arrays.asList(objectMapper.readValue(getClass().getResourceAsStream("/containers.json"), Container[].class)));

        productRepository.saveAll(Arrays.asList(objectMapper.readValue(getClass().getResourceAsStream("/products.json"), Product[].class)));

        passengerRepository.saveAll(Arrays.asList(objectMapper.readValue(getClass().getResourceAsStream("/passengers/passengers.json"), Passenger[].class)));

        stream(objectMapper.readValue(getClass().getResourceAsStream("/freight.json"), FreightDto[].class)).map(freightDto -> Freight.builder()
                        .id(freightDto.id())
                        .name(freightDto.name())
                        .type(freightDto.type())
                        .containers(freightDto.composition()
                                .stream()
                                .map(containerFullDto -> containerRepository.findById(containerFullDto.containerId())
                                        .orElse(null))
                                .filter(Objects::nonNull)
                                .collect(Collectors.toList()))
                        .supplier(companyRepository.findById(freightDto.supplierId())
                                .orElse(null))
                        .vendor(companyRepository.findById(freightDto.vendorId())
                                .orElse(null))
                        .build())
                .forEach(freightRepository::save);

        stream(objectMapper.readValue(getClass().getResourceAsStream("/train.json"), TrainDto[].class)).map(trainDto -> Train.builder()
                        .id(trainDto.id())
                        .name(trainDto.name())
                        .type(trainDto.type())
                        .carriages(trainDto.composition()
                                .stream()
                                .map(carriageFullDto -> carriageRepository.findById(carriageFullDto.carriageId())
                                        .orElse(null))
                                .filter(Objects::nonNull)
                                .collect(Collectors.toList()))
                        .supplier(companyRepository.findById(trainDto.supplierId())
                                .orElse(null))
                        .vendor(companyRepository.findById(trainDto.vendorId())
                                .orElse(null))
                        .build())
                .forEach(trainRepository::save);

        bridgeRepository.findAll()
                .forEach(bridge -> {
                    final Instant now = Instant.now();
                    final long millisToAdd = 10000;
                    IntStream.range(0, 200)
                            .boxed()
                            .map(integer -> BridgeOpeningTime.builder()
                                    .id((long) integer)
                                    .bridge(bridge)
                                    .openingTime(now.plusMillis(millisToAdd * integer * 2)
                                            .toEpochMilli())
                                    .closingTime(now.plusMillis(millisToAdd * (integer * 2 + 1))
                                            .toEpochMilli())
                                    .build())
                            .forEach(openingTimeRepository::save);
                });

    }

}

File: bl-central-server/bl-web-app/src/main/kotlin/org/jesperancinha/logistics/web/BridgeLogisticsInitializer.kt
package org.jesperancinha.logistics.web

import com.fasterxml.jackson.databind.ObjectMapper
import org.hibernate.Session
import org.jesperancinha.logistics.jpa.dao.*
import org.jesperancinha.logistics.web.dto.CarriageFullDto
import org.jesperancinha.logistics.web.dto.ContainerFullDto
import org.jesperancinha.logistics.web.dto.FreightDto
import org.jesperancinha.logistics.web.dto.TrainDto
import org.springframework.boot.CommandLineRunner
import org.springframework.context.annotation.Profile
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional
import java.time.Instant
import java.util.*
import java.util.function.Consumer
import java.util.stream.Collectors

@Component
@Profile("default", "local", "demo")
@Transactional
class BridgeLogisticsInitializer(
    private val bridgeRepository: BridgeRepository,
    private val carriageRepository: CarriageRepository,
    private val containerRepository: ContainerRepository,
    private val freightRepository: FreightRepository,
    private val productRepository: ProductRepository,
    private val trainRepository: TrainRepository,
    private val openingTimeRepository: OpeningTimeRepository,
    private val companyRepository: CompanyRepository,
    private val passengerRepository: PassengerRepository
) : CommandLineRunner {
    private val objectMapper = ObjectMapper()

    @Transactional
    override fun run(vararg args: String?) {
        bridgeRepository.deleteAll()
        companyRepository.deleteAll()
        carriageRepository.deleteAll()
        containerRepository.deleteAll()
        freightRepository.deleteAll()
        trainRepository.deleteAll()
        openingTimeRepository.deleteAll()
        bridgeRepository.saveAll(
            listOf(
                *objectMapper.readValue(
                    javaClass.getResourceAsStream("/bridges.json"),
                    Array<Bridge>::class.java
                )
            ).map { bridge: Bridge -> bridge.copy(id = null) }
        )
        companyRepository.saveAll(
            listOf(
                *objectMapper.readValue(
                    javaClass.getResourceAsStream("/companies.json"),
                    Array<Company>::class.java
                )
            )
        )
        carriageRepository.saveAll(
            listOf(
                *objectMapper.readValue(
                    javaClass.getResourceAsStream("/carriages.json"),
                    Array<Carriage>::class.java
                )
            ).map { carriage: Carriage -> carriage.copy(id = null) }
        )
        containerRepository.saveAll(
            listOf(
                *objectMapper.readValue(
                    javaClass.getResourceAsStream("/containers.json"),
                    Array<Container>::class.java
                )
            ).map { container: Container -> container.copy(id = null) }
        )
        productRepository.saveAll(
            listOf(
                *objectMapper.readValue(
                    javaClass.getResourceAsStream("/products.json"),
                    Array<Product>::class.java
                )
            ).map { product: Product -> product.copy(id = null) }
        )
        passengerRepository.saveAll(
            listOf(
                *objectMapper.readValue(
                    javaClass.getResourceAsStream("/passengers/passengers.json"),
                    Array<Passenger>::class.java
                )
            ).map { passenger: Passenger -> passenger.copy(id = null) }
        )
        objectMapper.readValue(
            javaClass.getResourceAsStream("/freight.json"),
            Array<FreightDto>::class.java
        ).map { freightDto: FreightDto ->
            Freight(
                id = freightDto.id,
                name = freightDto.name,
                type = freightDto.type,
                containers = freightDto.composition.mapNotNull { containerFullDto: ContainerFullDto ->
                    containerRepository.findById(containerFullDto.containerId)
                        .orElse(null)
                },
                supplier = companyRepository.findByIdOrNull(freightDto.supplierId),
                vendor =
                    companyRepository.findByIdOrNull(freightDto.vendorId)
            )
        }
            .map { freight: Freight -> freight.copy(id = null) }
            .forEach(freightRepository::save)

        objectMapper.readValue(javaClass.getResourceAsStream("/train.json"), Array<TrainDto>::class.java)
            .map { trainDto ->
                Train(
                    id = trainDto.id,
                    name = trainDto.name,
                    type = trainDto.type,
                    carriages = trainDto.composition
                        .stream()
                        .map { carriageFullDto: CarriageFullDto ->
                            carriageRepository.findById(carriageFullDto.carriageId)
                                .orElse(null)
                        }
                        .filter { obj: Carriage? -> Objects.nonNull(obj) }
                        .collect(Collectors.toList()),
                    supplier =
                        companyRepository.findByIdOrNull(trainDto.supplierId),
                    vendor =
                        companyRepository.findByIdOrNull(trainDto.vendorId),
                )
            }
            .map { train: Train -> train.copy(id = null) }
            .forEach(trainRepository::save)

        bridgeRepository.findAll()
            .forEach { bridge ->
                val now = Instant.now()
                val millisToAdd: Long = 10000
                (0..200)
                    .map { integer: Int ->
                        BridgeOpeningTime(
                            bridge = bridge,
                            openingTime =
                                now.plusMillis(millisToAdd * integer * 2)
                                    .toEpochMilli(),
                            closingTime =
                                now.plusMillis(millisToAdd * (integer * 2 + 1))
                                    .toEpochMilli()
                        )
                    }
                    .forEach(openingTimeRepository::save)
            }
    }
}